### ISSUE TITLE
question about camera matrixs for spatial coordinate conversion

### ISSUE THREAD

**Issue by hrz2000** [threestudio-project/threestudio#95]
How is the coordinate system of nerf defined in this code base? If I want to project a 3D point in space onto a 2D image, or combine the pixels on the 2D image with the depth map to lift it to 3D space, should I use a c2w matrix or an mvp_mtx? What are the internal and external parameters of a camera? Looking forward to an answer!

**Comment by bennyguo** [comment]
We adopt a right-hand coordinate system with: +x points outwards, +y points to the right, and +z points upwards. c2w defines how a point in camera coordinate system can be transformed to the world coordinate system. It contains information about camera rotation and translation. Use c2w if you want to project 2D points to 3D space using predicted depth. mvp_mtx is the projection matrix which defines how a 3D point is projected to the screen. It contains both information about camera extrinsics and intrinsics.

**Comment by hrz2000** [comment]
thanks for your quickly reply! i have some relevant promblem, Is the center coordinate of object is (0,0,0) in world coordinate system? i tried to project (0,0,0) to 2D image using mvp_mtx, however it did'nt show in the center of object, i'm confused of the concrete projection process, 
```python
def project_pts_on_img(points,
                raw_img,
                lidar2img_rt,
                color=None,
                thickness=-1):
    img = raw_img.copy()
    num_points = points.shape[0]
    pts_4d = np.concatenate([points[:, :3], np.ones((num_points, 1))], axis=-1)
    # import pdb;pdb.set_trace()
    pts_2d = pts_4d @ lidar2img_rt.T

    # cam_points is Tensor of Nx4 whose last column is 1
    # transform camera coordinate to image coordinate
    pts_2d[:, 2] = np.clip(pts_2d[:, 2], a_min=1e-5, a_max=99999)
    pts_2d[:, 0] /= pts_2d[:, 2]
    pts_2d[:, 1] /= pts_2d[:, 2]

    fov_inds = ((pts_2d[:, 0] < img.shape[1])
                & (pts_2d[:, 0] >= 0)
                & (pts_2d[:, 1] < img.shape[0])
                & (pts_2d[:, 1] >= 0))

    imgfov_pts_2d = pts_2d[fov_inds, :3]  # u, v, d

    for i in range(imgfov_pts_2d.shape[0]):
        depth = imgfov_pts_2d[i, 2]
        cv2.circle(
            img,
            center=(int(np.round(imgfov_pts_2d[i, 0])),
                    int(np.round(imgfov_pts_2d[i, 1]))),
            radius=1,
            color=color,
            thickness=thickness,
        )
    return img

```
the above is my projection code, i pass mvp_mtx as the lidar2img_rt, is it right?
for the 2d to 3d lift, my projection code is
```python
points = self.frustum.clone()# this is the piexel index
points[...,-1:] = depth
points = torch.cat((points[..., :2] * points[..., 2:3], points[..., 2:3]), -1)
points_3d = points @ c2w
```
could you please check my projection code for a right way of using?
thank you very much!

**Comment by bennyguo** [comment]
I think the `pts_2d` on this line is a homogenous:
```python
pts_2d = pts_4d @ lidar2img_rt.T
``` 

and you should do a perspective division to get normalized device coordinates (NDC) where x, y, z are in `[-1,1]`, after which you could shift and re-scale to get screen pixel coordinates:
```python
# x, y, z in pts_ndc are in [-1, 1]
pts_ndc = pts_2d / pts_2d[...,3:4]
```

To get 3D coordinates from pixels, a easy way would be to use `rays_o` and `rays_d` (here I assume the depth is the distance from the 3D points to the camera center):
```
pts_3d = rays_o + rays_d * depth
```

**Comment by voletiv** [comment]
In case it helps:
<img width="417" alt="Screenshot 2023-06-02 at 10 03 02" src="https://github.com/threestudio-project/threestudio/assets/22424247/d9405aeb-f0ad-4a8f-a942-850761b4616d">

**Comment by hrz2000** [comment]
Thank you for your kind answers. I will try it out

**Comment by claforte** [comment]
Closing since it sounds like this issue is resolved.

**Comment by sunshineatnoon** [comment]
Hello, Thanks for this awesome framework. May I ask what's the formulation of `mvp_mtx`? I see a typical `mvp_mtx` looks like
```
[[[ 0.0000,  2.7475,  0.0000,  0.0000], 
  [ 0.0000,  0.0000, -2.7475,  0.0000], 
  [-1.0020,  0.0000,  0.0000,  2.3048], 
  [-1.0000,  0.0000,  0.0000,  2.5000]]]
```
I'm interested in getting the focal length.

**Comment by sunshineatnoon** [comment]
Hi, I'm also interested in the camera matrices. For the intrinsic matrix, is the focal length defined as in [here](https://github.com/threestudio-project/threestudio/blob/main/threestudio/data/image.py#L126)? For the extrinsic matrix, is the `c2w` matrix defined as [R|T]? I'm trying to implement a new renderer here. Thanks
